data/com.luxoft.TemplateDBusService.xml

```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE node PUBLIC
+ "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
+ "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+  <interface name="com.luxoft.TemplateDBusService">
+    <method name="RemoveMeFoo">
+      <arg name="number" type="u" direction="in"/>
+      <arg name="is_odd" type="b" direction="out"/>
+    </method>
+
+    <signal name="RemoveMeBar">
+      <arg name="something" type="u"/>
+    </signal>
+
+    <property name="RemoveMeBaz" type="b" access="readwrite"/>
+  </interface>
```

Comment:
Is it possible to have clang-format rules for xml files also? These interface/signal/property names are part of the discussion regarding snake/camel case.

Reply:
This is how methods should be named in D-Bus. See the D-Bus documentation.

By "clang-format" here I assume you mean if there is a tool to lint D-Bus introspection XML, not that `clang-format` itself should be used. I do not know of any such tool. The DTD is there so I assume some tool can be used to verify that, but it would not help with the `name`attribute here, I am guessing.

Reply:
Oh it was the clang-tidy which contained the name formatting rules.
I was sort of assuming that there would be some naming convention already in place for the D-Bus methods.

Reply:
clang-tidy only checks C/C++ code, it knows nothing about D-Bus introspection XML. Probably not what you mean... but, hum... not sure I understand what is confusing here. To be extra clear.

- Convention in D-Bus is to name methods in CamelCase.
- gdbus-codegen-glibmm just takes the D-Bus method name in the introspection XML and generates a pure virtual method with that exact name in a C++ class (the stub).
- We subclass the stub and implement this pure virtual method when implementing a service.
- clang-tidy recognizes `override` and does not enforce the naming rules specified since... well... the method must be overridden... and comes from an external source (the code generated by the generator).

If we would have a hand writte C++ class with a pure virtual method, clang-tidy would have warned where that pure virtual method is specified, not where it is overridden.

Reply:
I meant that these same names are used as method names in dbus_service.cpp and there it makes it look odd, since it means not all method names use the same form.

Reply:
Yes, but the name comes from an external source. Nothing to do about it. Just to drive my point through... e.g. say that we want to, in a single project:

- Subclass a QWidget and override https://doc.qt.io/qt-5/qwidget.html#heightForWidth .
- Subclass a Glib::IOChannel and override http://manual.freeshell.org/glibmm-2.4/reference/html/classGlib_1_1IOChannel.html#abfb4faa07e80b86f21b0f18f0b8606cd
- Use gdbus-codegen-glibmm, as in this project, and implement a D-Bus service where D-Bus specifies that method names are CamelCase and the generator just extracts the name from the XML and does modify it in any way (as is done now).

Let's just say we do this in a single class for making the example simple. Would look something like this:

```C++
class Foo : <inherit from the stuff in the list above>
{
....
    // From QWidget. Uses lowerCaseCamelCase() style for methods.
    void hideEvent(QHideEvent *event) override;
....
    // From Glibmm. Uses snake_case() style for methods.
    Glib::IOStatus read_vfunc(char *buf, gsize count, gsize &bytes_read) override; 
...
    // From D-Bus. Uses CamelCase() style for methods.
    void RemoveMeFoo(MethodInvocation &invocation) override;
}
```

Three different styles, enforced on you from the outside world. It does not matter what style you prefer or enforce in other places of your project. These methods must be named the way they are since you `override`. You can not be consistent with all three of them at the same time.

And as I mentioned before, `clang-tidy` is smart here and does not warn about the identifier being wrongly named because it recognizes `override`.

Reply:
> Yes, but the name comes from an external source. Nothing to do about it. Just to drive my point home... e.g. say that we want to, in a single project:
> 

Then I am satisfied and know the reason for this difference in style. :)

> 
> And as I mentioned before, `clang-tidy` is smart here and does not warn about the identifier being wrongly named because it recognizes `override`.

Gottcha! :+1: